src/views/PositionPdfPlot/PositionPdfPlotWidget.tsx

```diff
@@ -0,0 +1,229 @@
+        segmentSize,
+        multiscaleFactor
+    }
+}
+
+const PositionPdfPlotWidget: FunctionComponent<Props> = ({fetchSegment, startTimeSec, endTimeSec, samplingFrequency, numPositions, segmentSize, multiscaleFactor, width, height}) => {
+    useRecordingSelectionTimeInitialization(startTimeSec, endTimeSec)
+    const { visibleTimeStartSeconds, visibleTimeEndSeconds } = useTimeRange()
+    const numTimepoints = Math.floor((endTimeSec - startTimeSec) * samplingFrequency)
+    const dataModel = usePositionPdfDataModel(fetchSegment, numTimepoints, numPositions, segmentSize, multiscaleFactor)
+
+    const {visibleValues, t1, t2} = useMemo(() => {
+        if (!visibleTimeStartSeconds) return {visibleValues: undefined, t1: 0, t2: 0}
+        if (!visibleTimeEndSeconds) return {visibleValues: undefined, t1: 0, t2: 0}
+        const i1 = Math.max(0, Math.floor((visibleTimeStartSeconds - startTimeSec) * samplingFrequency))
+        const i2 = Math.min(dataModel.numTimepoints, Math.ceil((visibleTimeEndSeconds - startTimeSec) * samplingFrequency))
+        if (i2 <= i1) return {visibleValues: undefined, t1: 0, t2: 0}
+        let downsampleFactor = 1
+        while ((i2 - i1) / downsampleFactor > 1000) {
+            downsampleFactor *= dataModel.multiscaleFactor
+        }
+        const j1 = Math.floor(i1 / downsampleFactor)
+        const j2 = Math.ceil(i2 / downsampleFactor)
+        const visibleValues = dataModel.get(j1, j2, downsampleFactor)
+        const t1 = startTimeSec + j1 * downsampleFactor / samplingFrequency
+        const t2 = startTimeSec + j2 * downsampleFactor / samplingFrequency
+        return {visibleValues, t1, t2}
+    }, [dataModel, visibleTimeStartSeconds, visibleTimeEndSeconds, startTimeSec, samplingFrequency])
+
+    const panelCount = 1
+    const { panelWidth, panelHeight } = usePanelDimensions(width, height, panelCount, panelSpacing, margins)
+    const pixelsPerSecond = usePixelsPerSecond(panelWidth, visibleTimeStartSeconds, visibleTimeEndSeconds)
+    const timeToPixelMatrix = use1dTimeToPixelMatrix(pixelsPerSecond, visibleTimeStartSeconds)
+
+    const pixelTimes = useMemo(() => {
+        const augmentedVisibleTimesMatrix = matrix([ [t1, t2], new Array(2).fill(1) ])
+        return multiply(timeToPixelMatrix, augmentedVisibleTimesMatrix).valueOf() as number[]
+    }, [t1, t2, timeToPixelMatrix])
+
+    const {minValue, maxValue} = useMemo(() => {
+        if (!visibleValues) return {minValue: 0, maxValue: 0}
+        return {
+            minValue: min(visibleValues.map(a => (min(a)))),
+            maxValue: max(visibleValues.map(a => (max(a)))),
+        }
+    }, [visibleValues])
+
+    const imageData = useMemo(() => {
+        if (!visibleValues) return undefined
+        if (minValue === undefined) return undefined
+        if (maxValue === undefined) return undefined
+        const N1 = visibleValues.length
+        const N2 = visibleValues[0].length
+        const imageData = new ImageData(N1, N2)
```

Comment:
What's an `ImageData`? It doesn't seem to be defined or imported anywhere.

Reply:
It's native data object that can be drawn to an HTML5 canvas directly -- sent directly to the GPU.